pulsar-broker/src/test/java/org/apache/pulsar/broker/delayed/InMemoryDeliveryTrackerTest.java

```diff
@@ -274,4 +274,69 @@ public void testDelaySequence(InMemoryDelayedDeliveryTracker tracker) throws Exc
         tracker.close();
     }
 
+    @Test(dataProvider = "delayedTracker")
+    public void testDelayedMessagesCountWithDuplicateEntryId(InMemoryDelayedDeliveryTracker tracker) throws Exception {
+        assertFalse(tracker.hasMessageAvailable());
+
+        // case1: addMessage() with duplicate entryId,
+        // getScheduledMessages() enter "cardinality <= n" and make tracker empty
+        assertTrue(tracker.addMessage(1, 1, 10));
+        assertTrue(tracker.addMessage(1, 2, 20));
+        assertTrue(tracker.addMessage(1, 2, 20));
+        assertTrue(tracker.addMessage(1, 3, 40));
+        assertTrue(tracker.addMessage(1, 4, 50));
+
+        clockTime.set(50);
+        assertTrue(tracker.hasMessageAvailable());
+        assertEquals(tracker.getNumberOfDelayedMessages(), 4L);
+        assertEquals(tracker.delayedMessageMap.size(), 4L);
+
+        Set<Position> scheduled = tracker.getScheduledMessages(10);
+        assertEquals(scheduled.size(), 4);
+        assertEquals(tracker.getNumberOfDelayedMessages(), 0L);
+
+
+
+        // case2: addMessage() with duplicate entryId,
```

Comment:
In case2 the comment says it enters cardinality > n, but with getScheduledMessages(10) and 4 unique entryIds it should hit the cardinality <= n branch. Could we adjust the comment to match the scenario (case3 seems to be the one exercising cardinality > n)?

Reply:
Yes. have changed the comment